core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java

```diff
@@ -817,17 +823,21 @@ public void asyncCache() throws Exception {
       PageId pageId = new PageId("5", i);
       assertTrue(mCacheManager.put(pageId, page(i, PAGE_SIZE_BYTES)));
     }
-    pageStore.setPutHanging(false);
-    //fallback to caller's thread when queue is full
+    // by setting the following line the hanging will only be stopped by the current thread
+    pageStore.setStopHangingThread(Thread.currentThread().getId());
+    // fallback to caller's thread (the current here) when queue is full
     assertTrue(mCacheManager.put(PAGE_ID1, PAGE1));
-    while (pageStore.getPuts() < threads) {
+    // Wait for all tasks to complete
+    // one for each thread worker thread, and one on the main thread
+    while (pageStore.getPuts() < threads + 1) {
```

Comment:
so the problem is all threads are hanging and the main thread is also hanging on put?
but why it can be passed locally and in some linux?

Reply:
There is a data race that causes the deadlock if not all async tasks complete before reaching line https://github.com/Alluxio/alluxio/pull/16731/files#diff-7a5716693110ccb17144045e526a6cf407af1ad7688f16e1e4481426689c51baR835
Then there will not be enough threads to run the tasks in the following lines.